File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/exception/GlobalExceptionController.java
package pl.madzierski.daniel.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;

@ControllerAdvice
public class GlobalExceptionController {

    @ExceptionHandler({AppRuntimeException.class})
    @ResponseStatus(HttpStatus.BAD_REQUEST)

    public ResponseEntity<String> handleAppRuntimeException(AppRuntimeException ex) {
        return ResponseEntity.status(ex.getType().getStatus()).body(ex.getType().getI18nMessage());
    }

    @ExceptionHandler({Exception.class})
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)

    public ResponseEntity<String> handleException(Exception ex) {
        ex.printStackTrace();
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error: " + ex.getMessage());
    }
}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/exception/GlobalExceptionController.kt
package pl.madzierski.daniel.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus


@ControllerAdvice
class GlobalExceptionController {


    @ExceptionHandler(AppRuntimeException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleAppRuntimeException(ex: AppRuntimeException) =
        ResponseEntity.status(ex.type.status).body(ex.type.i18nMessage)

    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    fun handleException(ex: Exception): ResponseEntity<String> {
        ex.printStackTrace()
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error: ${ex.message}")
    }

}